[TTS] Create script for processing TTS training audio

rlangman opened the pull request:
Signed-off-by: Ryan <rlangman@nvidia.com>

# What does this PR do ?

Add a script for doing audio preprocessing for TTS training.

**Collection**: [TTS]

# Changelog 
- Created a "data" directory under nemo.collections.tts following the same convention as the ASR & NLP code.
- Added data_util class for generic operations like processing training manifest. Not sure of volume normalization should be there, but currently have no better place to put it.
- Added silence trimming implementations using energy and VAD based approaches.
- Added audio_preprocessing script for running the data processing logic over a training manifest, with example config files.

# Usage

```
HYDRA_FULL_ERROR=1 python /home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/preprocess_audio.py \
    --config-path=/home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/config \
	--config-name=preprocessing.yaml \
	data_base_dir="/home/rlangman/Data/nemo_processing" \
	config.num_workers=2 \
	config.max_entries=100 \
	trim=vad \
	trim.device=cuda
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

# Additional Information

I manually tested the silence trimming using the openSLR Spanish data. For the 2 modes the dataset size decreases from:

Energy: 35 hours -> 28.8 hours
VAD: 35 hours -> 22.6 hours

This is because lots of the audio has noise in the long silences at the start and end of files. The VAD can differentiate the noise from speech while energy cannot.

Doing trimming on my workstation with the ~24k utterance dataset takes about 5 minutes with energy and 30 minutes with VAD.

More automated testing might be possible if we add CI tests that use actual audio, but I would leave that for a future PR as we do not yet have guidelines for TTS CI tests.

redoctopus reviewed (commented):
Thanks for adding this!

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+
+    # Iterate forwards over the utterance until we find the first frame_threshold consecutive speech frames.
+    start_i = None
+    for i in range(0, num_frames):
+        high_i = min(num_frames, i + frame_threshold)
+        if all(is_speech[i:high_i]):
+            start_i = i
+            break
+
+    # Iterate backwards over the utterance until we find the last frame_threshold consecutive speech frames.
+    end_i = None
+    for i in range(num_frames, 0, -1):
+        low_i = max(0, i - frame_threshold)
+        if all(is_speech[low_i:i]):
+            end_i = i
+            break
```
The min/max here might not make sense (I wrote it originally with the check sum() > threshold rather than all()). I think I should stop the iteration when the range becomes less than the threshold size.

redoctopus replied:
New version looks fine to me!

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+        return speech_frames
+
+    def _scale_sample_indices(self, start_sample_i: int, end_sample_i: int, sample_rate: int) -> Tuple[int, int]:
+        sample_rate_ratio = sample_rate / self.vad_sample_rate
+        start_sample_i = sample_rate_ratio * start_sample_i
+        end_sample_i = sample_rate_ratio * end_sample_i
```
pad_sample_indices() called after this is what handles the bounding and integer conversion. I could also bound the min/max and convert the final output to integer within this function too if it makes things clearer.

redoctopus replied:
Ah gotcha, I see it. Either is fine, though if it's kept as-is can you put in a note for the future in case anyone wants to call it by itself for whatever reason?

XuesongYang replied:
+1 with above suggestion.

rlangman commented on scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:
```diff
@@ -0,0 +1,19 @@
+  max_entries: 0
+  output_sample_rate: 0
+  volume_level: 0.95
+  min_duration: 0.5
+  max_duration: 10.0
```
In the end these are just examples that people will want to tweak depending on their dataset.

An upper bound of 10 seconds is a rule of thumb. I don't remember where I heard it originally, but I heard the same thing from Boris/Evelina months ago when we were talking about creating new HiFi-TTS like datasets. Its mainly done for computation reasons; the compute time and memory for a training batch is primarily dependent on the longest sequence in the batch. To use utterances much longer than 10 seconds you typically implement duration based batching (eg. there are a lot of papers that claim they use batch sizes of 10k or 20k audio frames, instead of fixed utterance count). A reason it comes up often is because TTS datasets are often derived from ASR datasets, and ASR likes training with utterances up to 20 seconds long.

The lower bound is more arbitrary. I would be OK with setting it to 0.1 for consistency. The reason I would set it higher than 0.1 here is to help identify/remove edge cases where the trimming ends up removing too much of the utterance. Alternatively I considered storing the original and trimmed durations in the output manifest and leaving the duration filtering/validation/analysis as a second step/script (the filtering could be done with a simple JQ command https://stedolan.github.io/jq/), but thought that was too complicated for an initial implementation.

redoctopus replied:
In general I'd be worried that people will blindly take whatever parameters are set in the example as defaults :P The reasoning is good and these seem fine, though.

XuesongYang replied:
thanks for explanation. how about make it consistent with other yamls, and leave a comment to user?

XuesongYang commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+        return speech_frames
+
+    def _scale_sample_indices(self, start_sample_i: int, end_sample_i: int, sample_rate: int) -> Tuple[int, int]:
+        sample_rate_ratio = sample_rate / self.vad_sample_rate
```
add `int()` to enforce return type hint as `Tuple[int, int]`. I would add try...except just in case zero-division.

rlangman replied:
Added positivity assertions on various input arguments, which also avoids zero-division

XuesongYang commented on scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:
```diff
@@ -0,0 +1,180 @@
+Most of these can also be done by the TTS data loader at training time, but doing them ahead of time
+lets us implement more complex processing, validate the corectness of the output, and save on compute time.
+
+$ HYDRA_FULL_ERROR=1 python /home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/preprocess_audio.py \
+    --config-path=/home/rlangman/Code/NeMo/scripts/dataset_processing/tts/audio_processing/config \
```
nit: `/home/rlangman/Code/NeMo` -> `<your_nemo_repo_root>`.

XuesongYang commented on scripts/dataset_processing/tts/audio_processing/config/preprocessing.yaml:
```diff
@@ -0,0 +1,19 @@
+name: "preprocessing"
+
+data_base_dir: "/home"
```
shall we leave it `???` or keep as-is?

XuesongYang commented on scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:
```diff
@@ -0,0 +1,180 @@
+    audio, sample_rate = librosa.load(input_path, sr=None)
+
+    if audio_trimmer is not None:
+        audio, start_i, end_i = audio_trimmer.trim_audio(audio=audio, sample_rate=sample_rate, audio_id=audio_filepath)
```
convert to string: `audio-id` is `str` but `audio_filepath` is `Path`.

XuesongYang replied:
convert to string: `audio-id` is `str` but `audio_filepath` is `Path`.

XuesongYang commented on scripts/dataset_processing/tts/audio_processing/preprocess_audio.py:
```diff
@@ -0,0 +1,180 @@
+
+    sf.write(file=output_path, data=audio, samplerate=sample_rate)
+
+    original_duration = librosa.get_duration(filename=audio_filepath)
```
convert to string: `filename` is `str` but `audio_filepath` is `Path`.

XuesongYang reviewed (commented):
LGTM. Added some nitpicks and discussion. 

It seems that this PR support either energy-based or vad-based trimming solution. For entergy-based method, the classic way is using two measures of energy and  zero-crossing rate (see ref https://web.ece.ucsb.edu/Faculty/Rabiner/ece259/Reprints/078_endpoint%20detection.pdf). We are good to ship energy only method, but we can add ZCR in future.

another thought, is it possible to make a hybrid solution with both energy and VAD?

redoctopus commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+                 consecutive audio frames classified as speech. Setting this value higher is more robust to
+                 false-positives (silence detected as speech), but setting it too high may result in very short
+                 speech segments being cut out from the audio.
+               frame_length: Length of audio frames to use when doing speech detection. This does not need to match
+                 the frame_length used any other part of the code or model.
+               frame_step: Stride of audio frames to use when doing speech detection. This does not need to match
+                 the frame_step used any other part of the code or model.
```
I'm in favor of that. It's pretty explicit here in the docstring that they can be different, but maybe we can call it `trim_win_length` and `trim_hop_length`?

rlangman commented:
> It seems that this PR support either energy-based or vad-based trimming solution. For entergy-based method, the classic way > is using two measures of energy and zero-crossing rate (see ref https://web.ece.ucsb.edu/Faculty/Rabiner/ece259/Reprints /078_endpoint%20detection.pdf). We are good to ship energy only method, but we can add ZCR in future.

> another thought, is it possible to make a hybrid solution with both energy and VAD?

The intention is to merge a prototype/skeleton implementation here and then improve on the algorithm(s) over time. ZCR could definitely be useful.

A hybrid solution with both energy and VAD also sounds like a good idea. I will need to think about how to refactor the code to support such a mode. For those 2 you can simply take the boolean speech frame array and do an AND with the two outputs, but that might not be as applicable for other approaches.

redoctopus commented:
@rlangman what's the status of this PR?

rlangman commented:
> @rlangman what's the status of this PR?

I will update the PR next week. In addition to addressing the PR comments, I talked with Fei and am seeing if I can make the VAD config more consistent with what it was trained on and potentially reuse some of the code from the VAD post-processor.

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+        # [num_frames, 2]
+        log_probs = self.vad_model(input_signal=audio_signal, input_signal_length=audio_signal_len)
+        probs = torch.softmax(log_probs, dim=-1)
+        probs = probs.cpu().detach().numpy()
```
Since VAD can run on GPU, transferring to CPU is needed to process it as numpy afterwards. Technically I might be able to transfer it after the next 2 lines.

I will swap the order of the detach() and cpu().

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+            pad_seconds=self.pad_seconds,
+        )
+
+        trimmed_audio = audio[start_i:end_i]
```
The +1 is implicitly inside the "get_start_and_end_of_speech" logic (as can be seen in the unit tests for the function). Its a bit hard to follow, but otherwise the +1 needs to be done at the frame level and before the padding rather than in this line at the sample level. I will see if I can make it clearer.

rlangman commented on nemo/collections/tts/data/data_utils.py:
```diff
@@ -0,0 +1,44 @@
+    if not (0.0 <= volume_level <= 1.0):
+        raise ValueError(f"Volume must be in range [0.0, 1.0], received {volume_level}")
+
+    return volume_level * (audio / np.max(np.abs(audio)))
```
I think all-zero audio is technically silence, so volume change should be a no-op (return audio unchanged).

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+    start_i = None
+    for i in range(0, num_frames):
+        high_i = min(num_frames, i + frame_threshold)
+        if all(is_speech[i:high_i]):
```
I think the existing logic works. It can be confusing to follow because for indexing in python the start_i is inclusive and end_i is exclusive.

If i = 1, and frame_threshold = 3, then high_i = 4. Then is_speech[1:4] returns speech frames 1, 2, and 3, which is the correct range.

I removed  the min/max and changed the range instead to avoid checking the boundaries with fewer than "speech_frame_threshold" values.

rlangman replied:
In place of this function there are also utilities in the VAD code (https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/utils/vad_utils.py#L643) which can take a similar input and produce a list with the boundaries for all speech segments. You can then find SoS and EoS by looking at the first and last entries. But I am inclined to try further large refactoring like that to a future PR after we have the general data processing template merged.

XuesongYang replied:
sounds good!

rlangman commented on nemo/collections/tts/data/audio_trimming.py:
```diff
@@ -0,0 +1,292 @@
+        end_sample_i = sample_rate_ratio * end_sample_i
+        return start_sample_i, end_sample_i
+
+    def trim_audio(self, audio: np.array, sample_rate: int, audio_id: str = "") -> Tuple[np.array, int, int]:
```
Should I add it to each implementation if the docstring is already on the AudioTrimmer(ABC) interface?

redoctopus reviewed (commented):
LGTM. @XuesongYang to check over his comments before approval

XuesongYang reviewed (approved):
LGTM. thanks.!
